x-pack/plugins/task_manager/task_store.ts

```diff
@@ -0,0 +1,370 @@
+   */
+  public async init() {
+    if (this.wasInitialized) {
+      return;
+    }
+
+    const properties = {
+      type: { type: 'keyword' },
+      task: {
+        properties: {
+          taskType: { type: 'keyword' },
+          runAt: { type: 'date' },
+          interval: { type: 'text' },
+          attempts: { type: 'integer' },
+          status: { type: 'keyword' },
+          params: { type: 'text' },
+          state: { type: 'text' },
+          user: { type: 'keyword' },
+          scope: { type: 'keyword' },
+        },
+      },
+    };
+
+    try {
+      const templateResult = await this.callCluster('indices.putTemplate', {
+        name: this.index,
+        body: {
+          index_patterns: [this.index],
+          mappings: {
+            _doc: {
+              dynamic: 'strict',
+              properties,
+            },
+          },
+          settings: {
+            number_of_shards: 1,
+            auto_expand_replicas: '0-1',
+          },
+        },
+      });
+      this.wasInitialized = true;
+      return templateResult;
+    } catch (err) {
+      throw err;
+    }
+
+    return;
+  }
+
+  /**
+   * Schedules a task.
+   *
+   * @param task - The task being scheduled.
+   */
+  public async schedule(taskInstance: TaskInstance): Promise<ConcreteTaskInstance> {
+    await this.init();
+    if (!this.supportedTypes.includes(taskInstance.taskType)) {
+      throw new Error(
+        `Unsupported task type "${
+          taskInstance.taskType
+        }". Supported types are ${this.supportedTypes.join(', ')}`
+      );
+    }
+
+    const { id, ...body } = rawSource(taskInstance);
+    const result = await this.callCluster('index', {
+      id,
+      body,
+      index: this.index,
+      type: DOC_TYPE,
+      refresh: true,
+    });
+
+    const { task } = body;
+    return {
+      ...taskInstance,
+      id: result._id,
+      version: result._version,
+      attempts: 0,
+      status: task.status,
+      runAt: task.runAt,
+      state: taskInstance.state || {},
+    };
+  }
+
+  /**
+   * Fetches a paginatable list of scheduled tasks.
+   *
+   * @param opts - The query options used to filter tasks
+   */
+  public async fetch(opts: FetchOpts = {}): Promise<FetchResult> {
+    const sort = paginatableSort(opts.sort);
+    return this.search({
+      sort,
+      search_after: opts.searchAfter,
+      query: opts.query,
+    });
+  }
+
+  /**
+   * Fetches tasks from the index, which are ready to be run.
+   * - runAt is now or past
+   * - id is not currently running in this instance of Kibana
+   * - has a type that is in our task definitions
+   *
+   * @param {TaskQuery} query
+   * @prop {string[]} types - Task types to be queried
+   * @prop {number} size - The number of task instances to retrieve
+   * @returns {Promise<ConcreteTaskInstance[]>}
+   */
+  public fetchAvailableTasks = async () => {
+    const { docs } = await this.search({
+      query: {
+        bool: {
+          must: [
```

Comment:
Do we also want this?

```suggestion
          must: [
            { term: { 'task.status': 'idle' } },
```

Reply:
Correction: no we do not want this. We want this method to fetch tasks of any status. It should be able to find running tasks and test them to see if they've been running longer than their timeout time.